aten/src/ATen/native/cuda/ReduceSumProdKernel.cu

```diff
@@ -154,6 +154,51 @@ struct prod_functor<c10::complex<at::Half>> {
 #endif
 };
 
+template <typename scalar_t, typename enable = void>
+struct xor_sum_functor {
+  void operator()(TensorIterator& iter) {
+    gpu_reduce_kernel<scalar_t, uint64_t>(
+        iter,
+        func_wrapper<uint64_t>(
+            [] GPU_LAMBDA(uint64_t a, uint64_t b) -> uint64_t {
+              return a ^ b;
+            }));
+  }
+};
+
+template <typename scalar_t>
+struct xor_sum_functor<scalar_t, std::enable_if_t<!std::is_integral_v<scalar_t>>> {
+  void operator()(TensorIterator& iter) {
+    gpu_reduce_kernel<scalar_t, double>(
+        iter,
+        // implicitly upcast scalar_t to double
+        func_wrapper<double>([] GPU_LAMBDA(double a, double b) -> double {
+          union {
+            double d;
+            uint64_t u;
+          } a_converter, b_converter, result_converter;
+
+          a_converter.d = a;
+          b_converter.d = b;
+          result_converter.u = a_converter.u ^ b_converter.u;
+          // return a double, otherwise uint64_t will be cast to double
+          // when accumulating and the result will be wrong
```

Comment:
Ho why?

Reply:
Lets say we have inputs 

a: 3.14159, a_bits: 4614256650576692846
b: 1.61803, b_bits: 4609965778477721196
a ^ b: 9219082337818812418

If we return this as int64_t, the next time it is used as an input, it will be cast to 9219082337818812418.0, which becomes 4890905006165143848 in bits, which would mess up the xor reduction here